File: packages/sdk-python/whyops/proxy.py
"""
Proxy helpers — patch existing openai / anthropic client instances
so all calls route through the WhyOps proxy.

No imports of openai or anthropic. Accepts whatever the user has.
"""
from __future__ import annotations

from typing import Any, TypeVar

from ._config import HEADER_AGENT_NAME, HEADER_ANTHROPIC_KEY

T = TypeVar("T")


def patch_openai(client: T, proxy_url: str, api_key: str, agent_name: str) -> T:
    """
    Patch an openai.OpenAI (or AsyncOpenAI) client in place.

    Sets base_url, api_key, and default_headers so all calls route
    through the WhyOps proxy automatically.

    Example::

        import openai
        from whyops import WhyOps

        sdk = WhyOps(api_key="...", agent_name="my-agent", agent_metadata={...})
        client = sdk.openai(openai.OpenAI(api_key="whyops_..."))
    """
    obj: Any = client
    obj.base_url = proxy_url.rstrip("/")
    obj.api_key = api_key

    existing = getattr(obj, "default_headers", {}) or {}
    obj.default_headers = {
        **existing,
        "Authorization": f"Bearer {api_key}",
        HEADER_AGENT_NAME: agent_name,
    }
    return client


def patch_anthropic(client: T, proxy_url: str, api_key: str, agent_name: str) -> T:
    """
    Patch an anthropic.Anthropic (or AsyncAnthropic) client in place.

    Example::

        import anthropic
        from whyops import WhyOps

        sdk = WhyOps(api_key="...", agent_name="my-agent", agent_metadata={...})
        client = sdk.anthropic(anthropic.Anthropic(api_key="whyops_..."))
    """
    obj: Any = client
    obj.base_url = proxy_url.rstrip("/")
    obj.api_key = api_key

    existing = getattr(obj, "default_headers", {}) or {}
    obj.default_headers = {
        **existing,
        HEADER_ANTHROPIC_KEY: api_key,
        HEADER_AGENT_NAME: agent_name,
    }
    return client

File: packages/sdk-typescript/src/proxy.ts
/**
 * Proxy helpers — patch existing OpenAI / Anthropic client instances
 * so all calls route through the WhyOps proxy with the correct headers.
 *
 * The SDK does NOT import openai or @anthropic-ai/sdk.
 * It accepts whatever client the user already has and patches it.
 */

import { HEADERS } from './config.js';

// ─── OpenAI ─────────────────────────────────────────────────────────────────

/**
 * Patches an OpenAI client instance to route through the WhyOps proxy.
 *
 * @param client   An `new OpenAI(...)` instance
 * @param proxyUrl WhyOps proxy base URL
 * @param apiKey   WhyOps API key
 * @param agentName Agent name for X-Agent-Name header
 * @returns The same client, mutated in place
 */
export function patchOpenAI<T extends OpenAILike>(
  client: T,
  proxyUrl: string,
  apiKey: string,
  agentName: string,
): T {
  // The OpenAI SDK exposes `baseURL` as a settable property
  (client as any).baseURL = proxyUrl;

  // Replace the internal API key with the WhyOps key
  (client as any).apiKey = apiKey;

  // Inject default headers
  const existing = (client as any).defaultHeaders ?? {};
  (client as any).defaultHeaders = {
    ...existing,
    'Authorization': `Bearer ${apiKey}`,
    [HEADERS.agentName]: agentName,
  };

  return client;
}

/**
 * Patches an Anthropic client instance.
 *
 * @param client   An `new Anthropic(...)` instance
 * @param proxyUrl WhyOps proxy base URL
 * @param apiKey   WhyOps API key
 * @param agentName Agent name for X-Agent-Name header
 * @returns The same client, mutated in place
 */
export function patchAnthropic<T extends AnthropicLike>(
  client: T,
  proxyUrl: string,
  apiKey: string,
  agentName: string,
): T {
  (client as any).baseURL = proxyUrl;
  (client as any).apiKey = apiKey;

  const existing = (client as any).defaultHeaders ?? {};
  (client as any).defaultHeaders = {
    ...existing,
    [HEADERS.apiKeyAnthropic]: apiKey,
    [HEADERS.agentName]: agentName,
  };

  return client;
}

// ─── Minimal structural types (not importing the actual SDKs) ────────────────

interface OpenAILike {
  baseURL: string;
  [key: string]: unknown;
}

interface AnthropicLike {
  baseURL: string;
  [key: string]: unknown;
}
